pydoctor/templatewriter/__init__.py

```diff
@@ -5,6 +5,364 @@
+        # If no meta pydoctor-template-version tag found, 
+        # it's most probably a placeholder template. 
+        version = -1
+        for meta in dom.getElementsByTagName("meta"):
+            if meta.getAttribute("name") != "pydoctor-template-version":
+                continue
+            if meta.hasAttribute("content"):
+                version_str = meta.getAttribute("content")
+                if version_str:
+                    try:
+                        version = int(version_str)
+                    except ValueError:
+                        warnings.warn(f"Could not read '{template_name}' template version: "
+                                "the 'content' attribute must be an integer")
+                else:
+                    warnings.warn(f"Could not read '{template_name}' template version: "
+                        f"the 'content' attribute is empty")
+            else:
+                warnings.warn(f"Could not read '{template_name}' template version: "
+                    f"the 'content' attribute is missing")
+        return version
+
+class TemplateLookup:
+    """
+    The L{TemplateLookup} handles the HTML template files locations. 
+    A little bit like C{mako.lookup.TemplateLookup} but more simple. 
+
+    The location of the files depends wether the users set a template directory 
+    with the option C{--template-dir}, custom files with matching names will be 
+    loaded if present. 
+
+    This object allow the customization of any templates, this can lead to warnings 
+    when upgrading pydoctor, then, please update your template.
+
+    @note: The HTML templates versions are independent of the pydoctor version
+           and are idependent from each other.
+
+    @see: L{Template}
+    """
+
+    _default_template_dir = 'templates'
+
+    def __init__(self) -> None:
+        """Init L{TemplateLookup} with templates in C{pydoctor/templates}"""
+        default_template_dir = Path(__file__).parent.parent.joinpath(self._default_template_dir)
+        self._templates: Dict[str, Template] = { t.name:t for t in (Template.fromfile(f) for f in 
+                default_template_dir.iterdir()) if t }
+
+        self._default_templates = copy.deepcopy(self._templates)
+
+
+    def add_template(self, template: Template) -> None:
+        """
+        Add a custom template to the lookup. 
+
+        Compare the passed Template version with default template, 
+        issue warnings if template are outdated.
+
+        Warns if the custom template is designed for an older version of pydoctor. 
+
+        @raises UnsupportedTemplateVersion: If the custom template is designed for a newer version of pydoctor. 
+        """
+        
+        try:
+            default_version = self._default_templates[template.name].version
+        except KeyError:
+            warnings.warn(f"Invalid template filename '{template.name}' (will be ignored). Valid filenames are: {list(self._templates)}")
```

Comment:
What does "will be ignored" mean exactly: that the template will be ignored or that the invalid name will be ignored? It seems ignoring the invalid name implemented, since the template object is stored in the `finally:` clause, but perhaps ignoring the template would be better.

Reply:
This warning takes care of situation where a user drops a `random.html` file inside their template folder and expect the file to appear in the output folder. We make it clear thet it doesn't work like that. 

What is your demand exactly ?

Reply:
I think it's a bit confusing that the template we're not going to use is stored in `self._templates` nevertheless. It is harmless, but it would be better to not store it, I think.

Reply:
Yes your're right